README.md
todo: replace time based crond script to restart usb-services with a watchdog function to restart service when usb process idles from /var/log as a pid

# REALLY,REALLY,REALLY-> (with VFS495) BUGGY **WIP

# to disable fprint with lightdm, xscreensaver, sudo begin

/etc/pam.d/...

these two TOP lines require a comment for each of these pam.d configuration files to disable fprint in each file
```
#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so
```




<br>


cat /etc/pam.d/system-auth
```
# Generated by authselect on Tue May 11 16:42:26 2021
# Do not modify this file manually.


#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so



.......
```

cat /etc/pam.d/lightdm
```
#%PAM-1.0

#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so




........
```

cat /etc/pam.d/lightdm-greeter 
```
#%PAM-1.0


#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so



...........
```

cat /etc/pam.d/lightdm-autologin 
```
#%PAM-1.0
#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so



...............
```

cat /etc/pam.d/sudo              
```
#%PAM-1.0


#auth		sufficient  	pam_fprint.so
#auth		sufficient  	pam_fprintd.so


........

```
cat /etc/pam.d/xscreensaver
```
#%PAM-1.0


#auth            sufficient      pam_fprint.so
#auth            sufficient      pam_fprintd.so

..............
```


# to disable fprint with lightdm, xscreensaver, sudo end





















# setenforce -> 0 (unless you train SELinux to catch all the permissions errors to allow lightdm, the sensor, dbus, /user/.fprint still working on a SELinux auto label to allow these fprint functions within "enforcing"

requires
```
pam_passwdqc-1.4.0-3.fc34.x86_64
gnome-keyring-pam-40.0-1.fc34.x86_64
systemd-pam-248.2-1.fc34.x86_64
pam-devel-1.5.1-5.fc34.x86_64
fprintd-pam-1.90.9-2.fc34.x86_64
pam_fprint-0.2-1.x86_64
pam-1.5.1-5.fc34.x86_64


authselect-libs-1.2.3-1.fc34.x86_64
authselect-1.2.3-1.fc34.x86_64
authselect-compat-1.2.3-1.fc34.x86_64

```
```
wget https://github.com/c4pt000/Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora/raw/master/pam.d.tar.gz-FPRINT-current-lightdm-sudo-xscreensaver.tar.gz
```
```
wget https://github.com/c4pt000/Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora/raw/master/authselect-FPRINT-current-lightdm-sudo-xscreensaver.tar.gz
```
/etc/pam.d
/etc/authselect

restart script for VFS495 sensor for crond
<br>
since sensor times out and has weird issues for VFS495 
<br>
so far with
<br>
Bus 001 Device 031: ID 138a:003f Validity Sensors, Inc. VFS495 Fingerprint Reader
<br>



* crond script to autorun restart for VFS495 sensor for fingerprint sensor to be more reactive to work
* script runs continuously every 5 minutes restarting the systemctl function to load the usb fingerprint module 
* since once the systemctl starts "once" the usb module can timeout and lose functionality causing issues where the usb module has to be
* reloaded hence this crond script to reload the usb module every 5 minutes,

```
export EDITOR=nano
crontab -e
```
```
@reboot                                /usr/bin/restart-fingerprint.sh &
*/15    *   *   *   *                                  /usr/bin/restart-fingerprint.sh &

```
<br>
cat /usr/bin/restart-fingerprint.sh 
<br>
chmod +x /usr/bin/restart-fingerprint.sh
<br>
/usr/bin/restart-fingerprint.sh &

```
#!/bin/bash
systemctl stop vcsFPServiceDaemon
systemctl daemon-reload
systemctl start vcsFPServiceDaemon

```



xscreensaver fprintd support
https://github.com/c4pt000/xscreensaver-fingerprint-atomic

to disable fingerprint functions in system as a regular user move the .fprint dir out of the way
<br>
$ mv /home/c4pt/.fprint /home/c4pt/.fprint.null
<br>
see restart-fingerprint.sh script








# Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora
revitalized fprint_demo-gui for fprintd-enroll gui (dinosaurs from the dead of fedora ~20)

![s1](https://raw.githubusercontent.com/c4pt000/Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora/master/finger-detect-image.png)
![s1](https://raw.githubusercontent.com/c4pt000/Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora/master/success-enroll.png)

# * dont use Validity-Sensor-Setup-4.5-136.0.x86_64.rpm creates more bugs?


then sh install.sh must install on openssl 1.1.1k



install.sh as root "chmod +x install.sh"

```
#!/bin/bash

yum reinstall redhat-lsb-core-4.1-52.fc34.x86_64 -y
yum install fprintd-pam-1.90.9-2.fc34.x86_64

cd /opt
git clone https://github.com/c4pt000/Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora
cd Validity-sensors-fprint-demo-fingerprint-gui-biometric-fedora
cd RPMS

export LC_ALL=C >> /root/.bashrc
source /root/.bashrc

ln -s /usr/lib64/libcrypto.so.1.1 /usr/lib64/libcrypto.so.0.9.8
ln -s /usr/lib64/libssl.so.1.1 /usr/lib64/libssl.so.0.9.8

rpm2cpio libfprint-0.0.6-2.x86_64.rpm | cpio -idmv /
ldconfig

rpm -Uvh pam_fprint-0.2-1.x86_64.rpm
ln -s /usr/lib/security/pam_fprint.so /usr/lib64/security/pam_fprint.so
cp -rf lib/security/pam_fprint.so /usr/lib64/
mkdir /usr/lib/security
cp -rf lib/security/pam_fprint.so /usr/lib/security/pam_fprint.so



rpm -Uvh --force Validity-Sensor-Setup-4.5-118.00.x86_64.rpm
rm -rf /etc/rc.d/init.d/vcsFPServiceDaemon 
cp -rf vcsFPServiceDaemon /etc/rc.d/init.d/vcsFPServiceDaemon
systemctl enable vcsFPServiceDaemon
systemctl start vcsFPServiceDaemon



echo "auth      sufficient pam_fprint.so" >>  /etc/pam.d/sudo
echo "auth      sufficient pam_fprintd.so" >>  /etc/pam.d/sudo

```

/etc/pam.d/sudo
---------------------
```
auth		sufficient  	pam_fprint.so
auth		sufficient  	pam_fprintd.so


auth       include      system-auth
account    include      system-auth
password   include      system-auth
session    optional     pam_keyinit.so revoke
session    required     pam_limits.so
session    include      system-auth

--------------------------------------





```


```
systemctl stop vcsFPServiceDaemon
systemctl start vcsFPServiceDaemon
```

restart script
<br>
nano /usr/bin/restart-fingerprint.sh
<br>
chmod +x /usr/bin/restart-fingerprint.sh
<br>

restart-fingerprint.sh
----------------------
```
systemctl daemon-reload
systemctl stop vcsFPServiceDaemon
systemctl start vcsFPServiceDaemon
```

command line enroll cli:
<br>
```
pam_fprint_enroll
```
gui based enroll:
```
fprint_demo
```

```
lsusb 
```
Bus 001 Device 006: ID 138a:003f Validity Sensors, Inc. VFS495 Fingerprint Reader
```
<br>
* allow regular user to access fingerprint sensor replace <Your User Name> with username of user

```
groupadd fingerprint
usermod -a -G fingerprint <Your User Name>
usermod -a -G input <Your User Name>


```
/etc/udev/rules.d/45-fingerprint.rules

your vendor id, device id
-----------------------------------

SUBSYSTEM=="usb_device", ACTION=="add",
SYSFS{idVendor}=="138a" ,
SYSFS{idProduct}=="003f",
SYMLINK+="fingerprint-%k",
GROUP="fingerprint",
MODE="666"

KERNEL=="uinput", MODE="666", GROUP="fingerprint"

```


<br>
<br>
<br>
<br>









# really buggy implementation

requires restarting vcsFPService

```
systemctl stop vcsFPServiceDaemon
systemctl start vcsFPServiceDaemon
```
cli:

pam_fprint_enroll

gui:
fprint_demo



```
rpm -qa | grep fprint && rpm -qa | grep Val

fprint-demo-0.4-2.x86_64
pam_fprint-0.2-1.x86_64
Validity-Sensor-Setup-4.5-118.00.x86_64
```



<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>


This project aims on reverse engineering protocol of Validity 138a:0090, 138a:0094, 138a:0097, 06cb:0081, 06cb:009a fingerprint readers, creating specification and FLOSS libfprint driver.

## Discussions

Main chat of this project: [![Gitter](https://img.shields.io/gitter/room/nwjs/nw.js.svg)](https://gitter.im/Validity90/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link)

libfrprint issue: [https://gitlab.freedesktop.org/libfprint/libfprint/issues/54](https://gitlab.freedesktop.org/libfprint/libfprint/issues/54)

Lenovo forums: [https://forums.lenovo.com/t5/Linux-Discussion/Validity-Fingerprint-Reader-Linux/td-p/3352145](https://forums.lenovo.com/t5/Linux-Discussion/Validity-Fingerprint-Reader-Linux/td-p/3352145)

## Notable files

1. [spec.md](spec.md) - Specification draft, the main work goes here right now.
2. [dissector.lua](dissector.lua) - Wireshark dissector for decrypting communication after key exchange.
3. ~[libfprint directory](libfprint) - libfprint repo with this driver integrated~ Not ready at the moment.
4. [prototype](prototype) - Standalone prototype(extremly ugly code, would be completly rewritten for driver)

## Testing

[Prototype](prototype) testers are needed, please share your result and join us in our [Gitter](https://gitter.im/Validity90/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link).

## Status

|   Device  | Status |
|-----------|--------|
| 138a:0090 | Scan, Image output, Internal DB check works in prototype. There is also a match-on-host [libfprint driver](https://github.com/3v1n0/libfprint) by @3v1n0 based on prototype source(138a:0090 only). You can use it untill proper unified driver is available. |
| 138a:0097 | Scan, Internal DB check works in prototype |
| 138a:0094 | Doesn't work yet, but looks promising. I will try to work on it after 90/97 |
| 06cb:0081 | So far looks identical to 138a:0094 |
| 06cb:009a | Init works, leds work, scan doesn't work yet |
| 138a:0091 | Different protocol, out of scope for this project. Check out [Validity91](https://github.com/hmaarrfk/Validity91) which aims to reverse engineer it.|


| 		      Task       			| Specification/Analysis  | Prototype   | Driver 	    |
|---------------------------|-------------------------|-------------|-------------|
| Initialization  		      | Done 					          | Done	 	    | Not Started |
| Configuration/Reconfig    | In progress 	          | In progress | Not Started |
| Pre TLS key exchange 	    | In progress 				    | Done        | Not Started |
| TLS 			                | Done 						        | Done  	    | Not Started |
| 90: Ops: scan, LED, etc| In progress  			      | Scan, LED works  | Not Started |
| 97: Ops: enroll, check, reset, LED, etc| In progress  			      | Check works  | Not Started |
| Image format  		        | In progress  			      | Done        | Not Started |


original code here
<br>
https://github.com/nmikhailov/Validity90
<br>
https://balintbanyasz.wordpress.com/2015/03/27/get-validity-vfs-495-fingerprint-reader-working-in-ubuntu-14-04/
<br>
https://forums.linuxmint.com/viewtopic.php?t=302947
<br>


pam_fprint-0.2/ChangeLog
commit d8349b68216108fb970414c8984a4e99b095a750
Author: Daniel Drake <dsd@gentoo.org>
Date:   Mon Nov 19 19:15:02 2007 +0000

    v0.2 release

commit a798ab6ac51d3e47aef658709e44d085a63d8ffe
Author: Vasily Khoruzhick <anarsoul@gmail.com>
Date:   Mon Nov 19 19:11:09 2007 +0000

    Add enrollment application

commit 74f7fc1439a9b1dff83f804fc2a62ca2ecd64e59
Author: Daniel Drake <dsd@gentoo.org>
Date:   Mon Nov 19 19:05:46 2007 +0000

    Don't try to serve sessions from remote hosts
    
    Avoids a remote box asking you to scan your fingerprint when you login
    over SSH.
    
    Based on a change made to pam_thinkfinger.

commit adb0ae4947091846bc2771c7a32e4c7ed93b5d33
Author: Daniel Drake <dsd@thief.(none)>
Date:   Thu Nov 15 10:41:59 2007 +0000

    v0.1 release

commit ae1f7800c5fbdbfb3a2945c28d984a2593cabd17
Author: Daniel Drake <dsd@gentoo.org>
Date:   Wed Oct 31 20:03:34 2007 +0000

    Initial commit


libfprint/libfprint-0.0.6/config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* Default visibility */
#define API_EXPORTED __attribute__((visibility("default")))

/* Debug message logging */
/* #undef ENABLE_DEBUG_LOGGING */

/* Message logging */
#define ENABLE_LOGGING 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* defined if XV video overlay is available */
/* #undef HAVE_XV */

/* Define to 1 if your C compiler doesn't accept -c and -o together. */
/* #undef NO_MINUS_C_MINUS_O */

/* Name of package */
#define PACKAGE "libfprint"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME "libfprint"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "libfprint 0.0.6"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "libfprint"

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.0.6"

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.0.6"

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif


libfprint/libfprint-0.0.6/libfprint/imgdev.c
/*
 * Core imaging device functions for libfprint
 * Copyright (C) 2007 Daniel Drake <dsd@gentoo.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <errno.h>

#include <glib.h>
#include <magick/ImageMagick.h>

#include "fp_internal.h"

static int img_dev_init(struct fp_dev *dev, unsigned long driver_data)
{
	struct fp_img_dev *imgdev = g_malloc0(sizeof(*imgdev));
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(dev->drv);
	int r = 0;

	imgdev->dev = dev;
	dev->priv = imgdev;
	dev->nr_enroll_stages = 1;

	/* for consistency in driver code, allow udev access through imgdev */
	imgdev->udev = dev->udev;

	if (imgdrv->init) {
		r = imgdrv->init(imgdev, driver_data);
		if (r)
			goto err;
	}

	return 0;
err:
	g_free(imgdev);
	return r;
}

static void img_dev_exit(struct fp_dev *dev)
{
	struct fp_img_dev *imgdev = dev->priv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(dev->drv);

	if (imgdrv->exit)
		imgdrv->exit(imgdev);

	g_free(imgdev);
}

int fpi_imgdev_get_img_width(struct fp_img_dev *imgdev)
{
	struct fp_driver *drv = imgdev->dev->drv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(drv);
	int width = imgdrv->img_width;

	if (width > 0 && imgdrv->enlarge_factor > 1)
		width *= imgdrv->enlarge_factor;
	else if (width == -1)
		width = 0;

	return width;
}

int fpi_imgdev_get_img_height(struct fp_img_dev *imgdev)
{
	struct fp_driver *drv = imgdev->dev->drv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(drv);
	int height = imgdrv->img_height;

	if (height > 0 && imgdrv->enlarge_factor > 1)
		height *= imgdrv->enlarge_factor;
	else if (height == -1)
		height = 0;

	return height;
}

static struct fp_img *im_resize(struct fp_img *img, unsigned int factor)
{
	Image *mimg;
	Image *resized;
	ExceptionInfo exception;
	MagickBooleanType ret;
	int new_width = img->width * factor;
	int new_height = img->height * factor;
	struct fp_img *newimg;

	/* It is possible to implement resizing using a simple algorithm, however
	 * we use ImageMagick because it applies some kind of smoothing to the
	 * result, which improves matching performances in my experiments. */

	if (!IsMagickInstantiated())
		InitializeMagick(NULL);
	
	GetExceptionInfo(&exception);
	mimg = ConstituteImage(img->width, img->height, "I", CharPixel, img->data,
		&exception);

	GetExceptionInfo(&exception);
	resized = ResizeImage(mimg, new_width, new_height, 0, 1.0, &exception);

	newimg = fpi_img_new(new_width * new_height);
	newimg->width = new_width;
	newimg->height = new_height;
	newimg->flags = img->flags;

	GetExceptionInfo(&exception);
	ret = ExportImagePixels(resized, 0, 0, new_width, new_height, "I",
		CharPixel, newimg->data, &exception);
	if (ret != MagickTrue) {
		fp_err("export failed");
		return NULL;
	}

	DestroyImage(mimg);
	DestroyImage(resized);

	return newimg;
}

int fpi_imgdev_capture(struct fp_img_dev *imgdev, int unconditional,
	struct fp_img **_img)
{
	struct fp_driver *drv = imgdev->dev->drv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(drv);
	struct fp_img *img;
	int r;

	if (!_img) {
		fp_err("no image pointer given");
		return -EINVAL;
	}

	if (!imgdrv->capture) {
		fp_err("img driver %s has no capture func", drv->name);
		return -ENOTSUP;
	}

	if (unconditional) {
		if (!(imgdrv->flags & FP_IMGDRV_SUPPORTS_UNCONDITIONAL_CAPTURE)) {
			fp_dbg("requested unconditional capture, but driver %s does not "
				"support it", drv->name);
			return -ENOTSUP;
		}
	}

	fp_dbg("%s will handle capture request", drv->name);

	if (!unconditional && imgdrv->await_finger_on) {
		r = imgdrv->await_finger_on(imgdev);
		if (r) {
			fp_err("await_finger_on failed with error %d", r);
			return r;
		}
	}

	r = imgdrv->capture(imgdev, unconditional, &img);
	if (r) {
		fp_err("capture failed with error %d", r);
		return r;
	}

	if (img == NULL) {
		fp_err("capture succeeded but no image returned?");
		return -ENODATA;
	}

	if (!unconditional && imgdrv->await_finger_off) {
		r = imgdrv->await_finger_off(imgdev);
		if (r) {
			fp_err("await_finger_off failed with error %d", r);
			fp_img_free(img);
			return r;
		}
	}

	if (imgdrv->img_width > 0) {
		img->width = imgdrv->img_width;
	} else if (img->width <= 0) {
		fp_err("no image width assigned");
		goto err;
	}

	if (imgdrv->img_height > 0) {
		img->height = imgdrv->img_height;
	} else if (img->height <= 0) {
		fp_err("no image height assigned");
		goto err;
	}

	if (!fpi_img_is_sane(img)) {
		fp_err("image is not sane!");
		goto err;
	}

	if (imgdrv->enlarge_factor > 1) {
		/* FIXME: enlarge_factor should not exist! instead, MINDTCT should
		 * actually look at the value of the pixels-per-mm parameter and
		 * figure out itself when the image needs to be treated as if it
		 * were bigger. */
		struct fp_img *tmp = im_resize(img, imgdrv->enlarge_factor);
		fp_img_free(img);
		img = tmp;
	}

	*_img = img;
	return 0;
err:
	fp_img_free(img);
	return -EIO;
}

int img_dev_enroll(struct fp_dev *dev, gboolean initial, int stage,
	struct fp_print_data **ret, struct fp_img **_img)
{
	struct fp_img *img = NULL;
	struct fp_img_dev *imgdev = dev->priv;
	struct fp_print_data *print;
	int r;

	/* FIXME: convert to 3-stage enroll mechanism, where we scan 3 prints,
	 * use NFIQ to pick the best one, and discard the others */

	r = fpi_imgdev_capture(imgdev, 0, &img);

	/* If we got an image, standardize it and return it even if the scan
	 * quality was too low for processing. */
	if (img)
		fp_img_standardize(img);
	if (_img)
		*_img = img;
	if (r)
		return r;

	r = fpi_img_to_print_data(imgdev, img, &print);
	if (r < 0)
		return r;
	if (img->minutiae->num < MIN_ACCEPTABLE_MINUTIAE) {
		fp_dbg("not enough minutiae, %d/%d", r, MIN_ACCEPTABLE_MINUTIAE);
		fp_print_data_free(print);
		return FP_ENROLL_RETRY;
	}

	*ret = print;
	return FP_ENROLL_COMPLETE;
}

static int img_dev_verify(struct fp_dev *dev,
	struct fp_print_data *enrolled_print, struct fp_img **_img)
{
	struct fp_img_dev *imgdev = dev->priv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(dev->drv);
	struct fp_img *img = NULL;
	struct fp_print_data *print;
	int match_score = imgdrv->bz3_threshold;
	int r;

	r = fpi_imgdev_capture(imgdev, 0, &img);

	/* If we got an image, standardize it and return it even if the scan
	 * quality was too low for processing. */
	if (img)
		fp_img_standardize(img);
	if (_img)
		*_img = img;
	if (r)
		return r;

	r = fpi_img_to_print_data(imgdev, img, &print);
	if (r < 0)
		return r;
	if (img->minutiae->num < MIN_ACCEPTABLE_MINUTIAE) {
		fp_dbg("not enough minutiae, %d/%d", r, MIN_ACCEPTABLE_MINUTIAE);
		fp_print_data_free(print);
		return FP_VERIFY_RETRY;
	}

	if (match_score == 0)
		match_score = BOZORTH3_DEFAULT_THRESHOLD;

	r = fpi_img_compare_print_data(enrolled_print, print);
	fp_print_data_free(print);
	if (r < 0)
		return r;
	if (r >= match_score)
		return FP_VERIFY_MATCH;
	else
		return FP_VERIFY_NO_MATCH;
}

static int img_dev_identify(struct fp_dev *dev,
	struct fp_print_data **print_gallery, size_t *match_offset,
	struct fp_img **_img)
{
	struct fp_img_dev *imgdev = dev->priv;
	struct fp_img_driver *imgdrv = fpi_driver_to_img_driver(dev->drv);
	struct fp_img *img = NULL;
	struct fp_print_data *print;
	int match_score = imgdrv->bz3_threshold;
	int r;

	r = fpi_imgdev_capture(imgdev, 0, &img);

	/* If we got an image, standardize it and return it even if the scan
	 * quality was too low for processing. */
	if (img)
		fp_img_standardize(img);
	if (_img)
		*_img = img;
	if (r)
		return r;

	r = fpi_img_to_print_data(imgdev, img, &print);
	if (r < 0)
		return r;
	if (img->minutiae->num < MIN_ACCEPTABLE_MINUTIAE) {
		fp_dbg("not enough minutiae, %d/%d", r, MIN_ACCEPTABLE_MINUTIAE);
		fp_print_data_free(print);
		return FP_VERIFY_RETRY;
	}

	if (match_score == 0)
		match_score = BOZORTH3_DEFAULT_THRESHOLD;

	r = fpi_img_compare_print_data_to_gallery(print, print_gallery,
		match_score, match_offset);
	fp_print_data_free(print);
	return r;
}

void fpi_img_driver_setup(struct fp_img_driver *idriver)
{
	idriver->driver.type = DRIVER_IMAGING;
	idriver->driver.init = img_dev_init;
	idriver->driver.exit = img_dev_exit;
	idriver->driver.enroll = img_dev_enroll;
	idriver->driver.verify = img_dev_verify;
	idriver->driver.identify = img_dev_identify;
}



libfprint/libfprint-0.0.6/libfprint/drivers/validity/vfsWrapper.h

/*! @file vfsWrapper.h
*******************************************************************************
* Helper functions for Validity driver interface functions
*
* This file contains the Helper functions for Validity driver interface functions
* and their definitions
*
* Copyright 2006 Validity Sensors, Inc. 

* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#ifndef __vfsWrapper_h__
#define __vfsWrapper_h__

int vfs_dev_init( void* pValidityDriver );

void vfs_dev_exit( void* pValidityDriver );

int vfs_capture( void* pValidityDriver,
   	            int initial);
int vfs_enroll( void* pValidityDriver,
   	            int initial,
   	            int stage,
	            unsigned char** ppFPrintData,
	            int *nFPrintDataSize );

int vfs_verify( void* pValidityDriver,
   	            unsigned char* pFPrintData,
	            int nFPrintDataSize );

int vfs_identify( void* pValDriver,
                  unsigned char **ppTemplatesData,
                  int *pTemplatesDataLengths,
                  int *pMatch_offset,
                  int nItems );

int vfs_get_img_width( void* pValidityContext );

int vfs_get_img_height( void* pValidityContext );

int vfs_get_img_datasize( void* pValidityContext );

unsigned char* vfs_get_img_data( void* pValidityContext );

void vfs_free_img_data( unsigned char* pImgData );

int vfs_get_matcher_type();

int vfs_set_matcher_type(int matcherType);  


typedef int (*PtrVfsDevInit)(void *);
typedef void (*PtrVfsCleanHandles)(void *);
typedef int (*PtrVfsWaitForService)(void);
typedef unsigned char* (*PtrVfsGetImgData)(void *);
typedef void (*PtrVfsFreeImgData)(unsigned char *);
typedef  int (*PtrVfsSetMatcherType)(int);
typedef int (*PtrVfsCapture)(void *, int);
typedef int (*PtrVfsGetImgDatasize) (void *);
typedef int (*PtrVfsGetImgWidth) (void *);
typedef int (*PtrVfsGetImgHeight) (void *);
typedef void (*PtrVfsDevExit)(void *);

#endif /*vfsWrapper */


libfprint/libfprint-0.0.6/libfprint/drivers/validity/vfsDriver.h

/*! @file vfsDriver.h
*******************************************************************************
* libfprint Interface Functions
*
* This file contains the libfprint interface functions and definitions
*
* Copyright 2006 Validity Sensors, Inc. 

* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#ifndef __vfsDriver_h__
#define __vcsDriver_h__

#define VCS_MAX_FINGER              10
#define VALIDITY_DRIVER_ID          10
#define VALIDITY_VENDOR_ID          0x138A
#define VALIDITY_PRODUCT_ID_301     0x0005  
#define VALIDITY_PRODUCT_ID_451     0x0007
#define VALIDITY_PRODUCT_ID_5111    0x0010
#define VALIDITY_PRODUCT_ID_5011    0x0011
#define VALIDITY_PRODUCT_ID_471	    0x003c
#define VALIDITY_PRODUCT_ID_5131    0x0018
#define VALIDITY_PRODUCT_ID_491	    0x003d
#define VALIDITY_PRODUCT_ID_495	    0x003f
#define VALIDITY_DRIVER_FULLNAME    "Validity Sensors"
#define VALIDITY_FP_COMPONENT       "Validity"
#define MAX_ENROLLMENT_PRINTS	    3
#define MAX_TEMPLATES		        10
#define VFS_NT_MATCHER	    	    1
#define VFS_COGENT_MATCHER	    2
#define VFS_FPRINT_MATCHER     3

#define VFS_RESULT_WRAPPER_OK 0

enum fp_capture_result
{
    FP_CAPTURE_ERROR = -1,
    FP_CAPTURE_COMPLETE = 1,
    FP_CAPTURE_FAIL
};

typedef struct validity_dev_s 
{
    void *pValidityCtx;
    void *hImage;
    void *pEnrollData;
}validity_dev;

#endif	__vfsDriver_h__ 
  

